cpp/models/abm/testing_strategy.cpp

```diff
@@ -105,12 +102,20 @@ void TestingScheme::update_activity_status(TimePoint t)
 
 bool TestingScheme::run_scheme(PersonalRandomNumberGenerator& rng, Person& person, TimePoint t) const
 {
-    if (t - person.get_time_of_last_test() > m_minimal_time_since_last_test) {
-        if (m_testing_criteria.evaluate(person, t)) {
-            double random = UniformDistribution<double>::get_instance()(rng);
-            if (random < m_probability) {
-                return !person.get_tested(rng, t, m_test_parameters);
-            }
+    auto test_result = person.get_test_result(m_test_parameters.type);
+    // If the agent has a test result valid until now, use the result directly
+    if ((test_result.type != TestType::Count) &&
+        (test_result.time_of_testing + m_test_parameters.validity_period >= t)) {
+        return test_result.is_allowed_to_enter;
+    }
+    if (m_testing_criteria.evaluate(person, t)) {
+        double random = UniformDistribution<double>::get_instance()(rng);
+        if (random < m_probability) {
+            // In this case, the time_of_testing in the past (i.e. the agent has already performed it).
+            bool is_person_allowed_to_enter = !person.get_tested(rng, t, m_test_parameters);
+            person.add_test_result(t - m_test_parameters.required_time, m_test_parameters.type,
+                                   is_person_allowed_to_enter);
```

Comment:
To reflect what is happening when a test is performed, this could be done in the `get_tested` method.

Reply:
Wouldn't it add more lines for each of the return in `get_tested`?

Reply:
Yes, it would. I guess it's fine here then. :)